File: library/src/main/java/io/github/anderscheow/validator/conditions/Condition.java
package io.github.anderscheow.validator.conditions;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import java.util.ArrayList;
import java.util.List;

import io.github.anderscheow.validator.rules.BaseRule;
import io.github.anderscheow.validator.util.ErrorMessage;
import io.github.anderscheow.validator.util.Validate;

public abstract class Condition extends ErrorMessage implements Validate {

    private List<BaseRule> baseRules;

    public Condition() {
        this.baseRules = new ArrayList<>();
    }

    public Condition(@StringRes int errorRes) {
        this.baseRules = new ArrayList<>();

        setErrorRes(errorRes);
    }

    public Condition(@NonNull String errorMessage) {
        this.baseRules = new ArrayList<>();

        setErrorMessage(errorMessage);
    }

    public Condition add(@NonNull BaseRule baseRule) {
        baseRules.add(baseRule);
        return this;
    }

    protected List<BaseRule> getBaseRules() {
        return baseRules;
    }
}

File: library/src/main/java/io/github/anderscheow/validator/conditions/Condition.kt
package io.github.anderscheow.validator.conditions

import androidx.annotation.StringRes
import io.github.anderscheow.validator.interfaces.ErrorImpl
import io.github.anderscheow.validator.interfaces.Validate
import io.github.anderscheow.validator.rules.Rule

abstract class Condition : ErrorImpl, Validate {

    val rules: List<Rule>

    constructor(rules: List<Rule>, @StringRes errorRes: Int) : super(errorRes) {
        this.rules = rules
    }

    constructor(rules: List<Rule>, errorString: String) : super(errorString) {
        this.rules = rules
    }

    constructor(rule: Rule, @StringRes errorRes: Int) : super(errorRes) {
        this.rules = listOf(rule)
    }

    constructor(rule: Rule, errorString: String) : super(errorString) {
        this.rules = listOf(rule)
    }
}

class ConditionBuilder {
    val conditionList = arrayListOf<Condition>()

    operator fun Condition.unaryPlus() {
        conditionList.add(this)
    }
}